File: dotnet-adapters-shortest-paths-impl-bsmock/com.programmerare.shortestpaths.adapter.bsmock/generics/PathFinderBsmockGenerics.java
package com.programmerare.shortestpaths.adapter.bsmock.generics;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import com.programmerare.shortestpaths.core.api.Vertex;
import com.programmerare.shortestpaths.core.api.Weight;
import com.programmerare.shortestpaths.core.api.generics.EdgeGenerics;
import com.programmerare.shortestpaths.core.api.generics.GraphGenerics;
import com.programmerare.shortestpaths.core.api.generics.PathFinderGenerics;
import com.programmerare.shortestpaths.core.api.generics.PathGenerics;
import com.programmerare.shortestpaths.core.impl.generics.PathFinderBase;
import com.programmerare.shortestpaths.core.pathfactories.PathFactory;

import edu.ufl.cise.bsmock.graph.ksp.Yen;

/**
 * "Adapter" implementation of the "Target" interface 
 * @author Tomas Johansson
 * @see https://en.wikipedia.org/wiki/Adapter_pattern
 */
public class PathFinderBsmockGenerics 
	< 
		P extends PathGenerics<E, V, W>,  
		E extends EdgeGenerics<V, W>, 
		V extends Vertex, 
		W extends Weight
	>
	extends PathFinderBase<P,  E, V, W> 
	implements PathFinderGenerics<P, E, V, W>
{
	private final edu.ufl.cise.bsmock.graph.Graph graphAdaptee;
	private final Yen yenAlgorithm;	
	
	protected PathFinderBsmockGenerics(
		final GraphGenerics<E, V, W> graph 
	) {
		this(
			graph, 
			null				
		);
	}
	protected PathFinderBsmockGenerics(
		final GraphGenerics<E, V, W> graph, 
		final PathFactory<P, E, V, W> pathFactory
	) {
		super(graph, pathFactory);
		this.yenAlgorithm = new Yen();
		this.graphAdaptee = new edu.ufl.cise.bsmock.graph.Graph();
		populateGraphAdapteeWithEdges();
	}
	
	private void populateGraphAdapteeWithEdges() {
		final List<E> edges = this.getGraph().getEdges();
		for (final E edge : edges) {
			this.graphAdaptee.addEdge(
				edge.getStartVertex().getVertexId(), 
				edge.getEndVertex().getVertexId(), 
				edge.getEdgeWeight().getWeightValue()
			);	
		}
	}

	@Override
	protected List<P> findShortestPathHook(
		final V startVertex, 
		final V endVertex, 
		final int maxNumberOfPaths
	) {
		final List<P> paths = new ArrayList<P>();
		
		final List<edu.ufl.cise.bsmock.graph.util.Path> pathList = yenAlgorithm.ksp(
			graphAdaptee, 
			startVertex.getVertexId(), 
			endVertex.getVertexId(), 
			maxNumberOfPaths
		);
		for (edu.ufl.cise.bsmock.graph.util.Path path : pathList) {
			final LinkedList<edu.ufl.cise.bsmock.graph.Edge> listOfEdges = path.getEdges();
			final List<E> edges = new ArrayList<E>();
			for (edu.ufl.cise.bsmock.graph.Edge edgeAdaptee : listOfEdges) {
				final E edge = getOriginalEdgeInstance(edgeAdaptee);
				edges.add(
					edge
				);				
			}
			final W totalWeight = super.createInstanceWithTotalWeight(path.getTotalCost(), edges);
			paths.add(super.createPath(totalWeight, edges));
		}
		return Collections.unmodifiableList(paths);
	}

	private E getOriginalEdgeInstance(final edu.ufl.cise.bsmock.graph.Edge edgeAdaptee) {
		return super.getOriginalEdgeInstance(edgeAdaptee.getFromNode(), edgeAdaptee.getToNode());
	}

}

File: dotnet-adapters-shortest-paths-impl-bsmock/com.programmerare.shortestpaths.adapter.bsmock/generics/PathFinderBsmockGenerics.cs
using edu.ufl.cise.bsmock.graph.ksp;
using com.programmerare.shortestpaths.core.api;
using com.programmerare.shortestpaths.core.impl.generics;
using com.programmerare.shortestpaths.core.pathfactories;
using com.programmerare.shortestpaths.core.api.generics;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace com.programmerare.shortestpaths.adapter.bsmock.generics
{
    /**
     * "Adapter" implementation of the "Target" interface 
     * @author Tomas Johansson
     * @see https://en.wikipedia.org/wiki/Adapter_pattern
     */
    public class PathFinderBsmockGenerics <P, E, V, W>
	    : PathFinderBase<P, E, V, W>
	    , PathFinderGenerics<P, E, V, W>

        where P : PathGenerics<E, V, W>
        where E : EdgeGenerics<V, W>
        where V : Vertex
        where W : Weight
    { 
	    private readonly edu.ufl.cise.bsmock.graph.Graph graphAdaptee;
	    private readonly Yen yenAlgorithm;	

	    public PathFinderBsmockGenerics(
		    GraphGenerics<E, V, W> graph 
	    ): 	this(
			    graph, 
			    null				
		    ) 
        {
	    }
	    protected PathFinderBsmockGenerics(
		    GraphGenerics<E, V, W> graph, 
		    PathFactory<P, E, V, W> pathFactory
	    ): base(graph, pathFactory) {
		    this.yenAlgorithm = new Yen();
		    this.graphAdaptee = new edu.ufl.cise.bsmock.graph.Graph();
		    PopulateGraphAdapteeWithEdges();
	    }
	
	    private void PopulateGraphAdapteeWithEdges() {
		    IList<E> edges = this.GetGraph().Edges;
		    foreach(E edge in edges) {
			    this.graphAdaptee.addEdge(
				    edge.StartVertex.VertexId, 
				    edge.EndVertex.VertexId, 
				    edge.EdgeWeight.WeightValue
			    );	
		    }
	    }

	    protected override IList<P> FindShortestPathHook(
		    V startVertex, 
		    V endVertex, 
		    int maxNumberOfPaths
	    ) {
            IList<P> paths = new System.Collections.Generic.List<P>();
		    IList<edu.ufl.cise.bsmock.graph.util.Path> pathList = yenAlgorithm.ksp(
			    graphAdaptee, 
			    startVertex.VertexId,
			    endVertex.VertexId,
			    maxNumberOfPaths
		    );
		    foreach(edu.ufl.cise.bsmock.graph.util.Path path in pathList) {
			    java.util.LinkedList<edu.ufl.cise.bsmock.graph.Edge> listOfEdges = path.getEdges();
			    IList<E> edges = new System.Collections.Generic.List<E>();
			    foreach(edu.ufl.cise.bsmock.graph.Edge edgeAdaptee in listOfEdges) {
				    E edge = GetOriginalEdgeInstance(edgeAdaptee);
				    edges.Add(
					    edge
				    );				
			    }
			    W totalWeight = base.CreateInstanceWithTotalWeight(path.getTotalCost(), edges);
			    paths.Add(base.CreatePath(totalWeight, edges));
		    }
            return new ReadOnlyCollection<P>(paths);
	    }

	    private E GetOriginalEdgeInstance(edu.ufl.cise.bsmock.graph.Edge edgeAdaptee) {
		    return base.GetOriginalEdgeInstance(edgeAdaptee.getFromNode(), edgeAdaptee.getToNode());
	    }
    }
}